contracts/contracts/strategies/sonic/SonicValidatorDelegator.sol

```diff
@@ -0,0 +1,318 @@
+
+        emit Undelegated(withdrawId, validatorId, undelegateAmount);
+    }
+
+    // slither-disable-start reentrancy-no-eth
+    function withdrawFromSFC(uint256 withdrawId)
+        external
+        onlyRegistrator
+        nonReentrant
+        returns (uint256 withdrawnAmount)
+    {
+        require(withdrawId < nextWithdrawId, "Invalid withdrawId");
+
+        // Can still withdraw even if the validator is no longer supported
+        // Load the withdrawal from storage into memory
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(!isWithdrawnFromSFC(withdrawId), "Already withdrawn");
+
+        uint256 sBalanceBefore = address(this).balance;
+
+        ISFC(sfc).withdraw(withdrawal.validatorId, withdrawId);
+
+        // Save state to storage
+        withdrawnAmount = address(this).balance - sBalanceBefore;
+        pendingWithdrawals -= withdrawal.undelegatedAmount;
+        withdrawals[withdrawId].undelegatedAmount = 0;
+
+        // Wrap Sonic (S) to Wrapped Sonic (wS)
+        IWrappedSonic(wrappedSonic).deposit{ value: withdrawnAmount }();
+
+        // Transfer the Wrapped Sonic (wS) to the Vault
+        // slither-disable-next-line unchecked-transfer unused-return
+        IERC20(wrappedSonic).transfer(vaultAddress, withdrawnAmount);
+
+        // withdrawal.undelegatedAmount & withdrawnAmount can differ in case of slashing
+        emit Withdrawn(
+            withdrawId,
+            withdrawal.validatorId,
+            withdrawal.undelegatedAmount,
+            withdrawnAmount
+        );
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /// @notice returns a bool whether a withdrawalId has already been withdrawn or not
+    function isWithdrawnFromSFC(uint256 withdrawId) public view returns (bool) {
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(withdrawal.validatorId > 0, "Invalid withdrawId");
+        return withdrawal.undelegatedAmount == 0;
+    }
+
+    /// @dev restake any pending validator rewards for all supported validators
+    function restakeRewards(uint256[] calldata validatorIds)
+        external
+        nonReentrant
+    {
+        uint256 totalRewards = 0;
+        for (uint256 i = 0; i < validatorIds.length; ++i) {
+            uint256 rewards = ISFC(sfc).pendingRewards(
+                address(this),
+                validatorIds[i]
+            );
+
+            if (rewards > 0) {
+                totalRewards += rewards;
+                ISFC(sfc).restakeRewards(validatorIds[i]);
+            }
+        }
+
+        // TODO use Deposit event or something else?
+        // The SFC contract will emit Delegated and RestakedRewards events
+        emit Deposit(wrappedSonic, address(0), totalRewards);
+    }
+
+    /**
+     * @notice To receive native S from SFC and Wrapped Sonic (wS)
+     */
+    receive() external payable {
+        require(
+            msg.sender == sfc || msg.sender == wrappedSonic,
```

Comment:
This is a nice check, but there is [this little devil](https://github.com/OriginProtocol/origin-dollar/pull/2335/files#diff-8da7c6574d9a346b9e74b8a9b465632eb613a64e1d1b64fca9488e45f23d6c15R32) lurking around :)

Reply:
I don't think that donating S to the contract would throw off the math: 
- We do the `address(this).balance` only in the `withdrawFromTheSFC` and we are looking at the increment that the withdrawal has caused... ignoring previous amount
- Also the checkBalance does ignore the `S` balance on the contract. 

The only complaint I have that the check might give a false sense of security, since anyone can call `withdrawTo` and send S to the contract